lib/executor/variable_arrival_rate.go

```diff
@@ -271,59 +170,74 @@ func (varc VariableArrivalRateConfig) HasWork(et *lib.ExecutionTuple) bool {
 //TODO: combine with the ConstantArrivalRate?
 type VariableArrivalRate struct {
 	*BaseExecutor
-	config             VariableArrivalRateConfig
-	plannedRateChanges []rateChange
+	config VariableArrivalRateConfig
 }
 
 // Make sure we implement the lib.Executor interface.
 var _ lib.Executor = &VariableArrivalRate{}
 
-// streamRateChanges is a helper method that emits rate change events at their
-// proper time.
-func (varr VariableArrivalRate) streamRateChanges(ctx context.Context, startTime time.Time) <-chan rateChange {
-	ch := make(chan rateChange)
-	go func() {
-		for _, step := range varr.plannedRateChanges {
-			offsetDiff := step.timeOffset - time.Since(startTime)
-			if offsetDiff > 0 { // wait until time of event arrives
-				select {
-				case <-ctx.Done():
-					return // exit if context is cancelled
-				case <-time.After(offsetDiff): //TODO: reuse a timer?
-					// do nothing
-				}
+func (varc VariableArrivalRateConfig) cal(et *lib.ExecutionTuple, ch chan<- time.Duration) {
+	// TODO: add inline comments with explanation of what is happening
+	// for now just link to https://github.com/loadimpact/k6/issues/1299#issuecomment-575661084
+	// TODO: rewrite this as state based functions/lambdas or something, drop the channel and let the user call
+	// it and precalculate the values, something like the `next` below
```

Comment:
If you don't end up with the state-machine implementation (and, having glanced at it, this seems simpler), can you remove these TODOs and move some of the comments from that PR to this one?

Reply:
resolved in d68ebe3 ?